### PR TITLE
[Non-record] PR #1855 + SWA blend — val_bpb 1.06140 (3-seed)

### PR DESCRIPTION
**Pull request by ishangodawatta**
## Summary

PR #1855's full stack with **Stochastic Weight Averaging blended into the EMA shadow before GPTQ**. Submitted as **non-record** (3-seed mean 1.06140 is +0.32 mbpb worse than current SOTA 1.06108) to contribute three findings useful to future submissions.

| seed | post-TTT val_bpb | artifact bytes | eval_time |
|------|------------------|----------------|-----------|
| 42   | 1.06089          | 15,905,575     | 527 s     |
| 0    | 1.06158          | 15,905,785     | 459 s     |
| 1234 | 1.06172          | 15,914,864     | 458 s     |
| **mean** | **1.06140**  | 15,908,741     | 481 s     |

3-seed std: 0.00044 BPB. All artifacts under 16 MB cap; all evals under 600 s cap.

### 1. Hardware-calibration shift (+1.25 mbpb)

A vanilla, byte-identical PR #1855 reproduction (Pod F, `SWA_ENABLED=0`, seed=42) on our 8×H100 SXM Runpod AP-IN-1 host produced **val_bpb 1.06114** vs PR #1855's published 1.05989 — a **+1.25 mbpb shift**. This confirms @harborglowvintage-oss / PR #1981's calibration-shift observation, but at much smaller magnitude than PR #1981's claimed ~5 mbpb.

| | val_bpb | artefact bytes | eval time |
|---|---|---|---|
| Pod F (vanilla, our hardware, seed=42) | 1.06114 | 15,906,766 | 532 s |
| PR #1855 published seed=42 | 1.05989 | 15,897,259 | 509 s |
| **Hardware shift** | **+1.25 mbpb** | +9,507 bytes | +23 s |

### 2. SWA composes cleanly with PR #1855's LQER + per-group pipeline

| | val_bpb | Δ vs vanilla on same hardware |
|---|---|---|
| Vanilla SOTA (extrapolated to our hardware) | ~1.06233 | — |
| Pod D6 (PR #1855 + SWA, 3-seed mean) | 1.06140 | **−0.93 mbpb** |

The SWA addition gives a real ~0.9 mbpb improvement on the same hardware, composes cleanly with LQER asymmetric int4 + per-group lrzip compression at SWA_BLEND=0.5 (the SWA contribution stays in the EMA-shadow blend rather than fully replacing it).

### 3. Reproducibility: 3-seed mean within ±0.4 mbpb per-seed of published

Per-seed Δ vs published (ours minus PR #1855):
- seed=42: +1.00 mbpb (our hardware shift)
- seed=0: +0.33 mbpb
- seed=1234: −0.37 mbpb

Variance shape similar to published (seed=42 < seed=0 < seed=1234 in our run; seed=42 < seed=0 < seed=1234 in published).

## What's the code change

**+23 lines** in `train_gpt.py`, gated by `SWA_ENABLED=1`. With `SWA_ENABLED=0` the script reproduces PR #1855 byte-for-byte. Implementation: in the final 5% of training, snapshot live weights every 50 steps; before EMA-applies, blend the SWA average into the EMA shadow at weight 0.5.

```python
# Inside train loop, after EMA update
if h.swa_enabled and step >= int(h.iterations * (1.0 - h.swa_window_frac)) and step % h.swa_interval == 0:
    with torch.no_grad():
        for name, t in _live_state.items():
            swa_sum[name].add_(t.detach().float())
        swa_count += 1

# Just before EMA-applies
if h.swa_enabled and swa_count > 0:
    log(f"swa: blending {swa_count} snapshots at weight {h.swa_blend}")
    for name in ema_state:
        swa_avg = swa_sum[name] / swa_count
        ema_state[name].mul_(1.0 - h.swa_blend).add_(swa_avg, alpha=h.swa_blend)
```

Defaults: `SWA_ENABLED=0`, `SWA_WINDOW_FRAC=0.05`, `SWA_INTERVAL=50`, `SWA_BLEND=0.5`.

## Reproduction

```bash
# 1. Pull precomputed CaseOps shards (saves ~30 min on-pod tokenisation)
huggingface-cli download romeerp/parameter-golf-caseops-v1 \
  --repo-type dataset --local-dir ./data

# 2. Install lrzip (required for COMPRESSOR=pergroup)
apt-get install -y lrzip

# 3. Train each seed on 8xH100 SXM, 600s wallclock
for SEED in 42 0 1234; do
  CASEOPS_ENABLED=1 \
  DATA_PATH=./data/datasets/datasets/fineweb10B_sp8192_lossless_caps_caseops_v1_reserved \
  TOKENIZER_PATH=./data/datasets/tokenizers/fineweb_8192_bpe_lossless_caps_caseops_v1_reserved.model \
  SWA_ENABLED=1 SWA_WINDOW_FRAC=0.05 SWA_INTERVAL=50 SWA_BLEND=0.5 \
  COMPRESSOR=pergroup \
  RUN_ID=E056_seed\${SEED} SEED=\${SEED} \
  MAX_WALLCLOCK_SECONDS=600 \
  torchrun --standalone --nproc_per_node=8 train_gpt.py
done
```

## Test plan

- [x] Trains within 600s wallclock on 8×H100 80GB SXM (4500 iterations achieved per seed)
- [x] All 3 artifacts under 16 MB cap (max: 15,914,864 bytes; mean: 15,908,741 bytes)
- [x] TTT eval completes within 600s eval cap (max: 527 s)
- [x] 3-seed mean reproduced; per-seed numbers attached as `train_seed{42,0,1234}.log`
- [x] `SWA_ENABLED=0` regression: Pod F replication of PR #1855 vanilla seed=42 within +1.25 mbpb of published
- [x] SWA snapshot/blend code reachable in trace (logs confirm `swa: blending N snapshots` line in each seed run)
- [x] pergroup compression confirmed working (artefact ~280 KB smaller than COMPRESSOR=brotli baseline)

## Why submitted as non-record

Our 3-seed mean (1.06140) is +0.32 mbpb worse than PR #1855's published 1.06108. This is within the hardware-calibration shift we measured (+1.25 mbpb) — the SWA addition is a real ~0.9 mbpb improvement on our hardware, but not enough to overcome the calibration shift and clear the 1.05608 record bar. Submitting as non-record so the SWA implementation, calibration data, and pergroup reproduction recipe are available to future submissions.

### CODE REVIEW

**Comment by harborglowvintage-oss** [comment]
Hi @IshanG97 — thanks for the citation! Small correction: the calibration-shift observation you referenced is actually from PR #1983 (my submission), not #1981 — that one belongs to @hectar-glitches. Happy to confirm the finding stands; just wanted the record to point to the right PR.

**Comment by ishangodawatta** [comment]
Thanks for the catch @harborglowvintage-oss — citation fixed in fb81710 (both occurrences in README: line 8 and line 33). PR #2049 now correctly attributes the calibration-shift finding to PR #1983. Apologies for the mix-up.
